### PR TITLE
feat(Optimizer): Add MSE Optimizer

### PR DESCRIPTION
**Pull request by shuokay**
The goal of this issue is to implement MSE Optimizer.
Finding the `rmin` and `rmax` through the MSE Optimizer:
1. Take the maximum and minimum values `[rmin, rmax]` as initial values, then
2. Iterate N times, each time attempting to narrow down the interval [rmin, rmax] to [rmin + eps, rmax - eps]. Calculate the MSE error between the fake_quantized tensor and the original tensor using `rmin` and `rmax`.
3.  Chose the `rmin` and `rmax` with the smallest MSE error as inputs for calculating `scale` and `zeropoint`.

implements https://github.com/huggingface/quanto/issues/135
# What does this PR do
1. Add `MseSymmetricOptimizer`
2. Add `_MseAffineOptimizer`, because `QModuleMixin` doesn't support optimizer for activations, I make it internal temporarily
4. Fix ruff deprecate parameter warnings

### CODE REVIEW

**Comment by shuokay** [comment]
@dacorvo This PR is ready to review

File: quanto/tensor/optimizers/mse_optimizer.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+
+class MseSymmetricOptimizer(SymmetricOptimizer):
+    def __init__(self, p: float = 2.0, iters: int = 80) -> None:
+        self.iters = iters
+        self.p = p
+        assert p >= 1.0
+        assert 0 < iters < 100
+
+    def optimize(self, base: Tensor, bits: int, axis: int) -> Tensor:
```
Comment (dacorvo) [inline quanto/tensor/optimizers/mse_optimizer.py:27]: You are using the AbsMax as a starting point, then iterate: please do it explicitly to avoid code duplication.
You can for instance inherit from AbsmaxOptimizer and call super().optimize, or simply instantiate an AbsmaxOptimizer in your code.

File: quanto/tensor/optimizers/mse_optimizer.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+import torch
+from torch import Tensor
+
+from quanto.tensor.core import axis_to_dim
+
+from .affine_optimizer import AffineOptimizer
+from .symmetric_optimizer import SymmetricOptimizer
+
+
+def _fake_quantize(base: Tensor, scale: Tensor, zeropoint: Union[int, Tensor], qmin: int, qmax: int) -> Tensor:
```
Comment (dacorvo) [inline quanto/tensor/optimizers/mse_optimizer.py:12]: Please reuse the actual quantizers to make usre the quantizing algorithms are identical.

File: quanto/tensor/optimizers/mse_optimizer.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+            score = (fq_base - base).abs().pow(self.p).mean(dim) if axis else (fq_base - base).abs().pow(self.p).mean()
+            if score.item() < best_score:
+                best_score = score.item()
+                best_rmax = new_rmax
+        scale = best_rmax / qmax
+        return scale
+
+
+# QModuleMixin didn't support optimizer for activation yet, we make this internal temporarily
+class _MseAffineOptimizer(AffineOptimizer):
```
Comment (dacorvo) [inline quanto/tensor/optimizers/mse_optimizer.py:46]: Optimizers are only used for weights quantization, and only really make a difference for bitwidth < 8.
In quanto, they will typically be used only to quantize to qint2 and qint4.
In other words, only MseAffineOptimizer is actually going to be used (if it is better than MaxOptimizer).

File: quanto/tensor/optimizers/mse_optimizer.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    def __init__(self, p: float = 2.0, iters: int = 80) -> None:
+        self.iters = iters
+        self.p = p
+        assert p >= 1.0
+        assert 0 < iters < 100
+
+    def optimize(self, base: Tensor, bits: int, axis: int) -> Tuple[Tensor, Tensor]:
+        best_score = float("inf")
+        dim = None if axis is None else axis_to_dim(base, axis)
```
Comment (dacorvo) [inline quanto/tensor/optimizers/mse_optimizer.py:55]: axis is never None, it is always 0 or -1

File: test/model/test_quantize_mlp.py [hunk 5]
```diff
@@ -154,12 +155,13 @@ def test_quantized_mlp_device_memory(weights, dtype, weights_only, device):
     b = io.BytesIO()
     torch.save(model.state_dict(), b)
     # Free device memory
+    model = None
```
Comment (dacorvo) [inline test/model/test_quantize_mlp.py:158]: Please remove this

File: test/model/test_quantize_mlp.py [hunk 6]
```diff
@@ -154,12 +155,13 @@ def test_quantized_mlp_device_memory(weights, dtype, weights_only, device):
...
     del model
     assert get_device_memory(device) == base_memory
     # Reload state dict on CPU
     b.seek(0)
     state_dict = torch.load(b, map_location=torch.device("cpu"), weights_only=weights_only)
-    assert get_device_memory(device) == 0
+    # assert get_device_memory(device) == 0
```
Comment (dacorvo) [inline test/model/test_quantize_mlp.py:164]: and this

File: test/model/test_quantize_mlp.py [hunk 7]
```diff
@@ -177,7 +179,9 @@ def test_quantized_mlp_device_memory(weights, dtype, weights_only, device):
 
 
 @pytest.mark.parametrize(
-    "weights, optimizer", [[qint8, AbsmaxOptimizer()], [qint4, MaxOptimizer()]], ids=["w-qint8", "w-qint4"]
+    "weights, optimizer",
+    [[qint8, AbsmaxOptimizer()], [qint4, MaxOptimizer()], [qint8, MseSymmetricOptimizer()]],
```
Comment (dacorvo) [inline test/model/test_quantize_mlp.py:183]: Add a test for MseAffine also

**Review by dacorvo — CHANGES_REQUESTED** [review]
Thank you for the pull-request. 
There is a misunderstanding however about the core design principles of quanto in general and the scope of optimizers.
quanto is lightweight and simple, allowing to combine the best quantization algorithms 
 by strictly following the pytorch extension patterns and staying modality agnostic.
Regarding the optimizers: they are only meant to be used to apply PTQ on weights.
For 8-bit quantization, a SymmetricOptimizer will be used (defaults to AbsmaxOptimizer), for lower bitwidth, an AffineOptimizer.
The AbsmaxOptimizer gives excellent results for 8-bit, so I don't think it is actually required to provide alternatives like the MseSymmetricOptimizer.
The MseAffineOptimizer might improve the performance of qint2 or qint4 models, but I will only merge it if you provide numbers proving it.

**Comment by shuokay** [comment]
> Thank you for the pull-request. There is a misunderstanding however about the core design principles of quanto in general and the scope of optimizers. quanto is lightweight and simple, allowing to combine the best quantization algorithms by strictly following the pytorch extension patterns and staying modality agnostic. Regarding the optimizers: they are only meant to be used to apply PTQ on weights. For 8-bit quantization, a SymmetricOptimizer will be used (defaults to AbsmaxOptimizer), for lower bitwidth, an AffineOptimizer. The AbsmaxOptimizer gives excellent results for 8-bit, so I don't think it is actually required to provide alternatives like the MseSymmetricOptimizer. The MseAffineOptimizer might improve the performance of qint2 or qint4 models, but I will only merge it if you provide numbers proving it.

@dacorvo We have some internal models that show better results using MSE compared to other methods, including AbsMax. However, I don't think I can publicly share these models 😂 I will return to this PR and finish it when I have a shareable number.
By the way, here are some numbers but they didn't compare to AbsMax. https://mqbench.readthedocs.io/en/latest/benchmark/ImageClassification/Benchmark.html
